### ISSUE TITLE
Add registry for customising permission policies

### ISSUE THREAD

**Issue by ARogovskyy** [wagtail/wagtail#6898]
### Is your proposal related to a problem?
Permissions for documents and images are typically managed through collections and/or `construct_document_chooser_queryset` or `construct_image_chooser_queryset` hooks. However, sometimes we need more fine-grained and customized control.

### Describe the solution you'd like

One example would be to restrict permissions for a user to see only the documents and images uploaded by themselves, and not just in choosers. While this can be done with groups and these hooks, a potentially easier solution would be to override `wagtail.documents.permissions.permission_policy` with a custom `PermissionPolicy`. This can be achieved with a hook that is called to retrieve/override the permission policy to use. The hook could be used like this:
```python
@hooks.register("register_documents_permission_policy")
def set_custom_permission_policy(old_policy):
    return CustomPermissionPolicy(...)
```

The only change needed would be to add the following lines to `permissions.py`:
```python
for fn in hooks.get_hooks('register_documents_permission_policy'):
    permission_policy = fn(permission_policy)
```

### Describe alternatives you've considered

The example above could be done with groups and the aforementioned hooks. However, that requires to
- register the hook as described in the [docs](https://docs.wagtail.io/en/stable/reference/hooks.html#construct-document-chooser-queryset)
- make a group for each user, and add the user to the group
- make a collection for each user
- give the group the right permissions
- automatically assign uploaded documents/images to the right collection

This is more work than just writing a custom `PermissionPolicy`. Additionally, the admin could change the groups of a user, which may be undersirable. The solution above does not need many changes and keeps the permission-related code in one place.

### Additional context

Overriding the permission policy would of course effectively disable the current collection based one. If implemented, the documentation should be adjusted to explain that one should inherit from `CollectionOwnershipPermissionPolicy` if the old behavior is still desired in some parts.

**Comment by davislx** [comment]
+1. Actually would like a solution to allow override permission policy in general.

I find myself wanting to write a permission policy for pages to allow only allow editors edit their own page. But current PermissionPolicy are all hard wired in wagtail.

@laymonage Can this be considered with the new PermissionPolicy refactoring?

**Comment by laymonage** [comment]
Thank you for bringing this up @davislx, I do plan on working more on the permission system in the next release. For pages, I think the first step would be to get #10702 in.

I'm not so sure about allowing permission policies to be customised for built-in models like pages, documents, and images because they're central to Wagtail and if you don't get the nitty-gritty details quite right it might result in undesired behaviours. However, I'll consider and discuss this with other core team members.

**Comment by davislx** [comment]
Hi @laymonage, thank you for the quick reply! 

We do use a custom page model, so if I read the code correctly, I think https://github.com/wagtail/wagtail/pull/10702 can solve our use case of only allowing editors to edit their own page. 
Double checking: Once that is released, we just need to override `permissions_for_user` for our page model, return a custom Page Permission Tester (that subclass the parent), and test for that page owner condition?

**Comment by cnk** [comment]
I was also hoping to override permission policies for core Wagtail models. I think this might be approachable now for Image and Document models but PagePermissionPolicy would be pretty tricky - not only because it is a complicated class but also because it is instantiated a lot of different places. That said, the 5.1 release was a big step forward in taming some of that complexity. 

I finally [wrote up the changes we made to implement our multitenant permission rules](https://cynthiakiser.com/blog/2023/11/05/permission-patches-for-multitenancy.html) in the new PagePermissionPolicy and CollectionOwnershipPermissionPolicy. The main challenge I had was that we use the request to infer the site and policies are often instantiated at class definition time rather than within a request cycle.

**Comment by laymonage** [comment]
Thanks @cnk, that is very helpful! And thank you for raising the concerns about `PagePermissionPolicy` being instantiated in different places. We fixed that in #11267 which was released in 6.0.

I explored the idea of adding a permission policy registry in #11268, but I wasn't sure whether we should be exposing the permission policy vs. something like the `PagePermissionTester` class. I was leaning towards the latter as it's closer to where specific actions are checked in the admin, but after doing some experiments I'm starting to have doubts. Perhaps it makes more sense for the permission tester to be part of the permission policy instead of the other way around, especially considering that not all models have (or need to have) tester classes currently.

I'll do some more discovery and report back 🙂
